contributor-docs/adrs/adr-020-live-regions.md

```diff
@@ -0,0 +1,165 @@
+- https://www.scottohara.me/blog/2022/02/05/are-we-live.html
+
+## Decision
+
+In order to have a common interop point for live region announcements, Primer React will
+make use of a `live-region` custom element from `@primer/live-region-element`.
+This package will be included and published from the `primer/react` repo.
+
+The custom element exposes a way to make announcements that can be used
+across frameworks. This makes it useful not only for Primer but GitHub as a
+whole.
+
+In addition, `@primer/react` will leverage and export the following helpers for
+use within Primer React and GitHub:
+
+- `announce()` and `announceFromElement()`, programmatic helpers for making
+  announcements
+- The `Status` component to correspond with `role="status"`
+- The `Alert` component to correspond with `role="alert"`
+
+Within `@primer/react`, we should lint against usage of `aria-live` and the
+corresponding roles (if possible) and suggest using these alternatives instead.
+
+### Impact
+
+This decision will impact existing usage of `aria-live`, `role="status"`, and
+`role="alert"` for the components listed below. These components will need to be
+updated to use the new approach, using one of the following approaches:
+
+- Use `announce()` or `announceFromElement()`
+- Use `Status` or `Alert`
+
+In addition, we should make sure that `<live-region>` is successfully included
+in GitHub.
+
+#### Instances of `aria-live`
+
+- InlineAutocomplete
+- InputValidation
+- SelectPanel
+- TreeView
+
+### Instances of `role="status"`
+
+- LiveRegion
+- TreeView
+- ActionList examples
+
+### Instances of `role="alert"`
+
+None
+
+### Instances of `LiveRegion`
+
+- DataTable
+
+## Examples
+
+### Announce on page load
+
+```tsx
+import {Status} from '@primer/react'
+
+function ExampleComponent() {
+  return <Status>Example page load message</Status>
+}
+```
+
+### Announce on content change
+
+```tsx
+import {Status} from '@primer/react'
+import {useState} from 'react'
+
+function ExampleComponent() {
+  const [count, setCount] = useState(0)
+  return (
+    <button
+      type="button"
+      onClick={() => {
+        setCount(count + 1)
+      }}
+    >
+      <Status>Count {count}</Status>
+    </button>
+  )
+}
+```
+
+### Announce programmatically
+
+```tsx
+import {announce} from '@primer/react'
+import {useState} from 'react'
+
+function ExampleComponent() {
+  const [results, setResults] = useState(data)
+
+  return (
+    <>
+      <input
+        type="text"
+        onChange={event => {
+          const filteredResults = data.filter(item => {
+            /* ... */
+          })
+          setResults(filteredResults)
+          announce(`${filteredResults.length} results available`)
+        }}
+      />
+      {/* ... */}
+    </>
+  )
+}
+```
+
+## Alternatives
+
+## Unresolved questions
+
+- What should happen if a component makes excessive announcements due to a
+  component that it renders? Is it possible to "group" announcements?
```

Comment:
I guess this one depends. Would we weigh this based on frequency of announcements, or repetitiveness? (e.g. announcing "loading" multiple times due to multiple spinners being rendered on the page at once).

Reply:
Great points, maybe both? Would be great to find a reference page that has this problem and see what we could do to make it obvious when something is firing a lot 😅 

Thankfully some of these announcements will just be dropped if they are announced quickly but would be great to highlight the fundamental problem that is going on here.

One option down the road could be to create "grouping" live regions that could batch up messages. Basically something like:

```tsx
<StatusGroup>
  <SomeComponent />
  <SomeComponent />
  <SomeComponent />
</StatusGroup>
```

Where `SomeComponent` renders `Status`. This grouping component could then collect messages and announce them together (or could expose an api to "summarize") but this is way down the line and may not even be useful 😂

Reply:
@owenniblock's API suggestion is interesting. Would the props be the same for when we don't expect a bunch of announcements to be rendered? Or is this a special API for when consumers think we might build up a bunch of announcements?

Reply:
@mperrotti I hadn't really thought that far forward, just had an idea and threw it out there 😅 

I guess those props would be optional and if there's no `announcement-id` then it assumes each item should be announced individually.

We could also split out the announcement into its own tag (something like):

```
<Announce  announcement-id="my-id" announcement-template="{n} workflows loading" announcement-frequency="2000">
   <Spinner />
</Announce>
```

Or maybe this API could form part of `<AriaStatus>` if that gains tractions.